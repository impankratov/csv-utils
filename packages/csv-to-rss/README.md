# csv-to-rss

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test csv-to-rss` to execute the unit tests via [Jest](https://jestjs.io).

## Running lint

Run `nx lint csv-to-rss` to execute the lint via [ESLint](https://eslint.org/).
