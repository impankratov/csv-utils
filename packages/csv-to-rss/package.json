{
  "name": "@im.pankratov/csv-to-rss",
  "version": "0.0.1"
}
