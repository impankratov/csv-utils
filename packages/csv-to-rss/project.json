{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/csv-to-rss/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/csv-to-rss/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/packages/csv-to-rss"],
      "options": {
        "jestConfig": "packages/csv-to-rss/jest.config.ts",
        "passWithNoTests": true
      }
    },
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/csv-to-rss",
        "tsConfig": "packages/csv-to-rss/tsconfig.lib.json",
        "packageJson": "packages/csv-to-rss/package.json",
        "main": "packages/csv-to-rss/src/index.ts",
        "assets": ["packages/csv-to-rss/*.md"]
      }
    }
  },
  "tags": []
}
