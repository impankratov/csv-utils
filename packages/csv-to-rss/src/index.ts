export * from './lib/csv-to-rss';
