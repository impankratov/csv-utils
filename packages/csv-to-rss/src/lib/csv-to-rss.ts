import { Options as CsvParseOptions, parse } from 'csv-parse';
import { createReadStream, createWriteStream } from 'node:fs';
import * as transform from 'parallel-transform';
import { pipeline } from 'stream';
import { promisify } from 'util';
import { parseToXmlFeed, XmlTransformOptions } from './parse-to-xml-feed';

const DEFAULT_PARALLEL = 1;

const pipelinePromisified = promisify(pipeline);

const parallelAsyncTransform = <T, U>(
  maxParallel: number,
  transformFn: (data: T) => Promise<U>
) =>
  transform(maxParallel, (data, callback) =>
    transformFn(data)
      .then((res) => callback(null, res))
      .catch(callback)
  );

export type CsvToRssTransformAsyncFn<Input, Output> = (
  v: Input
) => Promise<Output>;

export interface CsvToRssOptions<Input, Output> {
  parseOptions?: CsvParseOptions;
  transformFn: CsvToRssTransformAsyncFn<Input, Output>;
  parallel?: number;
  xmlTransformOptions: XmlTransformOptions
}

export function csvToRss<Input, Output>(
  options: CsvToRssOptions<Input, Output>,
  input: string,
  output: string
): Promise<void> {
  return pipelinePromisified(
    createReadStream(input),
    parse({ bom: true, columns: true, ...options.parseOptions }),
    parallelAsyncTransform(
      options.parallel || DEFAULT_PARALLEL,
      options.transformFn
    ),
    parseToXmlFeed(options.xmlTransformOptions),
    createWriteStream(output, {})
  );
}
