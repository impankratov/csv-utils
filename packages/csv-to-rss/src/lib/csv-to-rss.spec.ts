import { createReadStream, createWriteStream } from 'node:fs';
import { csvToRss } from './csv-to-rss';

jest.mock('node:fs', () => {
  const originalModule = jest.requireActual('node:fs');

  const { Readable, Writable } = require('node:stream');

  const createReadStreamMock = jest.fn();

  let output: string[] = [];
  (createReadStreamMock as any).__mockOutput = (o: string[]) => (output = o);

  createReadStreamMock.mockImplementation(() => {
    const readable = new Readable();

    output.forEach((l: string) => readable.push(l));

    readable.push(null);

    return readable;
  });

  class WriteInMemory extends Writable {
    buffer: string = '';

    constructor() {
      super();
    }

    _write(chunk: Buffer, _: any, next: () => any) {
      this.buffer += chunk;
      next();
    }

    reset() {
      this.buffer = '';
    }
  }

  const MockWriteStream = new WriteInMemory();

  const createWriteStreamMock = jest.fn().mockImplementation(() => {
    MockWriteStream.reset();
    return MockWriteStream;
  });

  (createWriteStreamMock as any).__mock = MockWriteStream;

  return {
    __esModule: true,
    ...originalModule,
    createReadStream: createReadStreamMock,
    createWriteStream: createWriteStreamMock,
  };
});

const records = [
  {
    Id: '0001',
    Date: '07/05/1999',
    Title: ' Lorem ipsum dolor sit amet',
    Description:
      'Lorem ipsum dolor sit amet qui minim labore adipisicing minim sint cillum sint consectetur cupidatat.',
  },
  {
    Id: '0002',
    Date: '15/12/2000',
    Title: 'Lorem ipsum',
    Description:
      'Lorem ipsum dolor sit amet qui minim labore adipisicing minim sint cillum sint consectetur cupidatat.',
  },
  {
    Id: '0003',
    Date: '21/03/2003',
    Title: 'Lorem',
    Description:
      'Lorem ipsum dolor sit amet qui minim labore adipisicing minim sint cillum sint consectetur cupidatat.',
  },
];

const objectToCsv = (obj: Object) =>
  Object.values(obj).reduce(
    (res, v, i, arr) => `${res},${v}${arr.length - i === 1 ? '\n' : ''}`
  );

const getDummyCsvData = () =>
  [
    // 'Id,Title,Description,Date\n',
    `${Object.keys(records[0]).join(',')}\n`,
  ].concat(records.map(objectToCsv));

describe('csvToRss', () => {
  it('should work', async () => {
    (createReadStream as any).__mockOutput(getDummyCsvData());

    const TEST_INPUT = 'input.csv';
    const TEST_OUTPUT = 'output.xml';
    const TRANSFORM_FN = jest.fn((v) => Promise.resolve(v));

    await csvToRss(
      {
        transformFn: TRANSFORM_FN,
        xmlTransformOptions: {
          objectName: 'entry',
        },
      },
      TEST_INPUT,
      TEST_OUTPUT
    );

    expect(createReadStream).toHaveBeenCalledWith(TEST_INPUT);
    expect(TRANSFORM_FN.mock.calls).toMatchSnapshot();

    expect(createWriteStream).toHaveBeenCalledWith(TEST_OUTPUT, {});
    expect((createWriteStream as any).__mock.buffer).toMatchSnapshot();
  });
});
