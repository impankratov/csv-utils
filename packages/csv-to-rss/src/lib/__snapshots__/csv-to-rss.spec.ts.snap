// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csvToRss should work 1`] = `
Array [
  Array [
    Object {
      "Date": "07/05/1999",
      "Description": "Lorem ipsum dolor sit amet qui minim labore adipisicing minim sint cillum sint consectetur cupidatat.",
      "Id": "0001",
      "Title": " Lorem ipsum dolor sit amet",
    },
  ],
  Array [
    Object {
      "Date": "15/12/2000",
      "Description": "Lorem ipsum dolor sit amet qui minim labore adipisicing minim sint cillum sint consectetur cupidatat.",
      "Id": "0002",
      "Title": "Lorem ipsum",
    },
  ],
  Array [
    Object {
      "Date": "21/03/2003",
      "Description": "Lorem ipsum dolor sit amet qui minim labore adipisicing minim sint cillum sint consectetur cupidatat.",
      "Id": "0003",
      "Title": "Lorem",
    },
  ],
]
`;

exports[`csvToRss should work 2`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<entry>
  <Id>0001</Id>
  <Date>07/05/1999</Date>
  <Title> Lorem ipsum dolor sit amet</Title>
  <Description>Lorem ipsum dolor sit amet qui minim labore adipisicing minim sint cillum sint consectetur cupidatat.</Description>
</entry>
<entry>
  <Id>0002</Id>
  <Date>15/12/2000</Date>
  <Title>Lorem ipsum</Title>
  <Description>Lorem ipsum dolor sit amet qui minim labore adipisicing minim sint cillum sint consectetur cupidatat.</Description>
</entry>
<entry>
  <Id>0003</Id>
  <Date>21/03/2003</Date>
  <Title>Lorem</Title>
  <Description>Lorem ipsum dolor sit amet qui minim labore adipisicing minim sint cillum sint consectetur cupidatat.</Description>
</entry>
"
`;
